RHOAIENG-22480: Adding Guardrails Orchestrator use case examples

skrthomas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
https://issues.redhat.com/browse/RHOAIENG-22480
## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

skrthomas commented:
Screen Recording 2025-03-31 at 4.49.07 PM.mov

skrthomas commented:
![image](https://github.com/user-attachments/assets/a62225af-78b3-4e14-ab2d-62ee226cb384)

skrthomas commented:
![image](https://github.com/user-attachments/assets/549e4991-ca7d-4b24-a22d-8341e61a3c4a)

skrthomas commented:
![image](https://github.com/user-attachments/assets/393b401c-2665-47d3-9247-642ecb1a7d37)

skrthomas commented:
![image](https://github.com/user-attachments/assets/d574c7d7-ad50-49d6-bf51-dbe373e04727)

skrthomas commented:
Compare VSCode rendering:
![image](https://github.com/user-attachments/assets/02c682d5-c204-4b8c-969c-6c90caed79e3)

skrthomas commented:
![image](https://github.com/user-attachments/assets/a6e7737a-b0c1-4ff5-bc79-98ea3f51689c)

skrthomas commented:
![image](https://github.com/user-attachments/assets/b2c1dfb2-5105-4265-bc47-6ebeb9bcac37)

skrthomas commented:
![image](https://github.com/user-attachments/assets/56b930ff-eabe-4ba1-a211-e1b8fb87ab30)

BSynRedhat commented on assemblies/configuring-the-guardrails-orchestrator-service.adoc:
```diff
@@ -31,13 +31,14 @@ The following sections describe how to do the following tasks:
 
 include::modules/deploying-the-guardrails-orchestrator-service.adoc[leveloffset=+1]
 include::modules/guardrails-orchestrator-parameters.adoc[leveloffset=+1]
+include::modules/guardrails-orchestrator-scenarios.adoc[]
```
<img width="1112" alt="Screenshot 2025-04-04 at 11 41 19" src="https://github.com/user-attachments/assets/6a225a58-f28f-4f52-b54a-3f172245a1a1" />
I think we will need to update the introduction module of the section as well to reference these examples.

BSynRedhat commented on modules/guardrails-orchestrator-scenarios.adoc:
```diff
@@ -0,0 +1,218 @@
+
+[role='_abstract']
+
+The following example demonstrates how to use Guardrails Orchestrator to monitor user inputs to your LLM, specifically to protect against hateful and profane language (HAP) and personally identifiable information (PII). Comparison queries without detectors enabled are shown to illustrate the differences in responses when guardrails are disabled versus enabled. 
```
I thought this was going to be laid out more like 3 individual examples, with a common prerequisite section.  I would start with the 1) 'without detectors' section, then 2) HAP example, and then 3) PII example. I would make a bullet list of these 3 in the abstract section too. So something like: 
The following example demonstrates how to use Guardrails Orchestrator in 3 example scenarios:
* Using the Guardrails Orchestrator without detectors, to demonstrate its use without specific detectors enabled.
* Monitoring user inputs to your LLM to protect against hateful and profane language (HAP) using the HAP detector.
* Monitoring user inputs to your LLM to guard against the inclusion of personally identifiable information (PII) using the PII detector.

skrthomas replied:
In Slack today, Christina said that she tested the PII detector and couldn't get it working, so just to put the HAP detector example.

BSynRedhat commented on modules/guardrails-orchestrator-scenarios.adoc:
```diff
@@ -0,0 +1,218 @@
+* You have cluster administrator privileges for your OpenShift cluster.
+* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
+* You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
+
```
I think these prerequisites are complete, but @christinaexyou may confirm?

BSynRedhat commented on modules/guardrails-orchestrator-scenarios.adoc:
```diff
@@ -0,0 +1,218 @@
+.Procedure
+
+. Define a `ConfigMap` object in a YAML file to specify HAP and PII detectors. For example, create a file named `orchestrator_cm.yaml` with the following content:
++
```
So, for this section, I would recommend re-ordering it the way I outlined in my comment on 
 above -
putting the 'an example of a query not running the PII detector:' and 'an example of a query not running the HAP detector:' sections in one general 'no detectors enabled' example', 
then following that, a 2) HAP detector in use 
and then 3) a PII detector.  
@christinaexyou does this make sense?

skrthomas replied:
To have a "no detectors enabled" section, we would put a config map without detectors and then show the queries. As I said in my other comment, I think the contrasting query is more helpful in the procedure as a direct contrast to the query with detectors enabled. That way, you can see as you go through the procedure, oh this is what I would get if the HAP detector wasn't enabled. You won't have to scroll up to a separate procedure to check then scroll back down to the "enabled" version. Just how I am thinking about the user flow but @christinaexyou feel free to correct or suggest a different handling of this.

BSynRedhat replied:
Hmm, ok well my logic for the suggestion of a separated-out layout is the comment @christinaexyou made in the gdoc 2.19 TAI Tasks:
 " _Suggestion from Christina: “Have 3 use-cases 1) sending a request straight to the LLM (i.e. without detectors activating), then 2) HAP detector: sending a query that activates that  endpoint, and 3) PII detector: sending a query that activates this endpoint.” (Please note: I may not have noted the above down entirely accurately - B)_"
 
Can you send me a link to the source material @christinaexyou supplied so that I can have a look at that? That might help me understand the layout options. Is it on the TrustyAI website yet?
Otherwise, I think we should ask our team for guidance, as I am confused about your layout from a docs perspective.

skrthomas replied:
The source material is from the same working doc that we're sharing: https://docs.google.com/document/d/1VP58VOoJVGxpwTis1nB9fFhk2GBfjk0zXwv2gB7wStc/edit?tab=t.px0e08jhdrd. Notice that the example of "without" query is sequenced directly after the example query "with". The only source I have is that I pinged Christina to ask about why she removed the PII from the configmap supplied and she said that the PII detector wasn't working with her tests, so we could just use HAP.

BSynRedhat reviewed (changes requested):
I've made a few comments about and tagged Christina as well :)

skrthomas commented on modules/guardrails-orchestrator-standalone-detection-scenario.adoc:
```diff
@@ -0,0 +1,32 @@
+.Prerequisites
+* You have completed Steps 1-3 of the "Monitoring user inputs with the Guardrails Orchestrator service" example. 
+
+. Query to detect instances of HAP:
```
@christinaexyou I'm unsure if this standalone detection should be part of the other HAP scenario or if its better to leave it as separate but with the prerequisites to do the configmap for HAP?

christinaexyou replied:
I think you can put this right after the other HAP scenario with a brief explanation along the lines of, "You can also enable standalone detections on text by querying the `api/v2/text/detection/content` endpoint."

skrthomas replied:
Ok, I moved this under the HAP as an optional final step.

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+ config.yaml: |
+   chat_generation:
+     service:
+       hostname: llm-predictor.guardrails-test.svc.cluster.local
```
Add footnote explaining that the hostname here should point to the LLM service that we want to guardrail against

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+   detectors:
+     hap:
+       type: text_contents
+       service: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local
```
Add footnote explaining that the hostname here should point to the HAP detector service

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+
+.Procedure
+
+. Define a `ConfigMap` object in a YAML file to specify an HAP detector. For example, create a file named `orchestrator_cm.yaml` with the following content:
```
Rephrase to the first sentence: "Define a `ConfigMap` object in a YAML file to specify the LLM service you wish to guardrail against and the HAP detector service you want to run guardrailing with."

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+----
+GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <USER_NAMESPACE>)
+----
+
```
I think it makes more sense to add lines 88-111 (where we show the LLM generated outputs without the HAP detector enabled) before we enable the HAP detector

skrthomas replied:
I can see it both ways. Do you think we should leave this as a bullet step, like contextual information, or actually proceduralize it, telling customers to run the query without the HAP detector enabled?

skrthomas replied:
Based on your comment here, I'm going to assume you think this should be an actual step we tell users to take: https://github.com/opendatahub-io/opendatahub-documentation/pull/708#discussion_r2031551952 but if you think it should be a point of comparison, I can change it back to a bullet.

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <USER_NAMESPACE>)
+----
+
+. Query the model, enabling the HAP detector:
```
To be more specific, rephrase to "Query the orchestrator's `api/v2/chat/completions-detections` endpoint while enabling the enabling the HAP detector to generate a guardrailed response."

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+{"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0. ...
+----
++
+* For contrast, here is an example of a query not running the HAP detector:
```
Rephrase to "Query the orchestrator's api/v2/chat/completions-detections endpoint without the HAP detector enabled to generate a response without guardrails."

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
+----
+{"id":"cmpl-f6da55d06ade4792a33d4ae67a07cc38","object":"chat.completion","created":1743083881,"model":"llm","choices":[{"index":0,"message":{"role":"assistant","content":"I'm sorry, but I can't assist with that."},"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":56,"total_tokens":69,"completion_tokens":13}}
+----
+
```
Add conclusion here, "As demonstrated, when we enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs will be clearly flagged and model outputs will not be generated."

skrthomas replied:
Ok I like that context. Can I add something similar under the "without guardrails", like "As demonstrated, when HAP detections are not enabled on model inputs and outputs via the Guardrails Orchestrator, inappropriate inputs are generated when prompted. "

christinaexyou replied:
I like that idea but if you look at the output response, it is _not inapproppriate_. I would reword as "When HAP detections are not enabled on model inputs and outputs via the Guardrails Orchestrator, the model continues to generate outputs as normal."

skrthomas commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,111 @@
++
+[source,terminal]
+----
+GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <USER_NAMESPACE>)
```
Good catch, thank you!

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,137 @@
+       default_threshold: 0.5
+
+----
+<1> The `hostname` value should specify the LLM service that we want to guardrail against.
```
Replace hostname with `chat_generation.hostname`

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,137 @@
+       hostname: llm-predictor.guardrails-test.svc.cluster.local <1>
+       port: 8080
+   detectors:
+     hap:
```
This part should be:
```
  hap:
        type: text_contents
        service:
          hostname: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local
          port: 8000
```

skrthomas replied:
Ok addressed here. Is the spacing OK here: https://github.com/opendatahub-io/opendatahub-documentation/pull/708/files#diff-ceeb06c2cc384aa4d2fb9ec167be5b7f1c4a6ada0bd6c0bc3ef5c3d904346ba3R34-R38

skrthomas replied:
Weird now I am seeing the reply comment box, but basically I keep one commit in my PRs by using `git commend` to amend commits, so I don't have to squash at the end. You can still see changes made by clicking "Compare" after the force push in the PR dialogue.

christinaexyou commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,137 @@
++
+As demonstrated, when you enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs are clearly flagged and model outputs are not generated.
+
+. *Optional:* You can also enable standalone detections on text by querying the `api/v2/text/detection/content` endpoint by using the following query:
```
Delete ` by using the following query` to be more concise

skrthomas commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,137 @@
+.Example response
+[source]
+----
+{"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0. ...
```
Ah sorry, good catch I think I got this mixed up because the message response for without enabled is "I'm sorry, but I can't assist with that." and then the text with it enabled is "<explicit_text>, I really hate this stuff" but then further down its the message "Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed."...I think when I was moving stuff around I was thinking "oh the guardrailed response wouldn't have the "explicit_text". But I see its the _message_ that indicates the guardrail.

skrthomas commented:
@sheltoncyril can you PTAL at this from QE perspective?

sheltoncyril commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+
+* You have cluster administrator privileges for your OpenShift cluster.
+* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
+* You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
```
Would it be possible make `"Deploying the Guardrails Orchestrator"` a clickable link to the actual page?

skrthomas replied:
This is on the same page, just above this topic, its sorta hard to see with the screencaptured preview, but its only a topic away via scroll, and when things are on the same page, we don't typically add hyperlinks:
![image](https://github.com/user-attachments/assets/f714a3ea-0c3c-4802-9476-215b22a3795d)

sheltoncyril commented:
/lgtm

eturner24 commented on assemblies/configuring-the-guardrails-orchestrator-service.adoc:
```diff
@@ -30,19 +30,14 @@ It is underpinned by the open-source project link:https://github.com/foundation-
-* Monitor user-inputs to your LLM using this service
-// * Use the Guardrails Orchestrator service 
+* Monitor user-inputs to your LLM
+* *Optional* Enable the OpenTelemetry exporter for and observing metrics and tracing
```
```suggestion
* *Optional* Enable the OpenTelemetry exporter for observing metrics and tracing
```

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+
+* You have cluster administrator privileges for your OpenShift cluster.
+* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
+* You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
```
```suggestion
* You have deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
```

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+       default_threshold: 0.5
+
+----
+<1> The `chat_generation.service.hostname` value should specify the LLM service that we want to guardrail against.
```
```suggestion
<1> Specifies the LLM service to guardrail against.
```
Simplified a bit since we have the callout right there with the value

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+
+----
+<1> The `chat_generation.service.hostname` value should specify the LLM service that we want to guardrail against.
+<2> The `hap.service.hostname` value should specify the name of the HAP detector service. 
```
```suggestion
<2> Specifies the name of the HAP detector service. 
```
This feels slightly redundant because the value is literally called `hap.service.hostname` - is there something else that they should know about this or some reason to call this value out specifically?

skrthomas replied:
this is just how we write things in OCP docs so we can have a complete sentence in a callout with a subject and verb 🤷‍♀️

skrthomas replied:
I think the rationale for writing like that, from OCP peer review thinking, is otherwise, just starting with "Specifies" is a fragment.

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+{"id":"cmpl-f6da55d06ade4792a33d4ae67a07cc38","object":"chat.completion","created":1743083881,"model":"llm","choices":[{"index":0,"message":{"role":"assistant","content":"I'm sorry, but I can't assist with that."},"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":56,"total_tokens":69,"completion_tokens":13}}
+----
++
+As demonstrated, when HAP detections are not enabled on model inputs and outputs through the Guardrails Orchestrator, the model continues to generate outputs as normal. 
```
I don't follow what "generating outputs as normal" is. Is there a more specific term we can use here? "...the model's output is not filtered." ?

eturner24 replied:
I also don't see the same sample explicit text input that the second example has. Does this one need it? (I don't completely understand the code sample but I'm assuming if we wanted to show a comparison of two example responses we would want to use the same input/query.)
> [{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff",

skrthomas replied:
@christinaexyou Just want to double check with you because the example response provided in the doc does not have the same explicit text input: https://docs.google.com/document/d/1VP58VOoJVGxpwTis1nB9fFhk2GBfjk0zXwv2gB7wStc/edit?tab=t.px0e08jhdrd. Should the "You <explicit_text>, I really hate this stuff" be in this response too? 

@eturner24 for your first question, I think maybe we can clarify to "the model generates output without flagging unsuitable inputs." but @christinaexyou may clarify this.

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+{"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0. ...
+----
++
+As demonstrated, when you enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs are clearly flagged and model outputs are not generated.
```
```suggestion
When you enable HAP detections on model inputs and outputs using the Guardrails Orchestrator, unsuitable inputs are flagged and model outputs are not generated.
```

eturner24 commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+ "detectors": {
+   "hap": {}
+ },
+ "content": "You <explicit_text>, I really hate this stuff"
+}'
+----
++
+.Example response
+----
+{"detections":[{"start":0,"end":36,"text":"You <explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0.9634239077568054}]}
```
I like that I can match up the content in the example with the output. I don't see that same parallelism in the other code samples. Is it possible to do this?

skrthomas replied:
Sorry, are you meaning the other Guardrails code examples? Can you point me to the exact ones you're talking about? I'm not sure if you're referring to ones in this PR or on the Guardrails docs generally, but maybe these can be addressed in a follow-up task since this is for 2.19? @eturner24

eturner24 replied:
Yes, we can definitely leave this for 2.20! I can discuss with you offline as well - might be easier to explain that way. :D

BSynRedhat replied:
Sara has told me that Christina said that 'the output responses are OK' here, so there is no further action needed on Em's comment.

skrthomas commented:
Thanks for the review @eturner24 ! I addressed your feedback and looped in @christinaexyou for the one question I had on the model response. I put what was provided to me in the sample code, so I'm not sure if there's any additional response information that may be missing.

skrthomas commented:
@BSynRedhat Can you PTAnotherL at this and lmk if you have feedback?

skrthomas commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+
+* You have cluster administrator privileges for your OpenShift cluster.
+* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
+* You have deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
```
When I was talking to @eturner24  about this, we discussed that this topic is on the same page, like a scroll up, so the link isn't necessary like it would be if it were on a separate page. I can add the link though if its a RHOAI standard. I'm getting mixed signals on what's the right thing to do.

skrthomas commented:
> How was this addition added? I only see one commit overall ?

Pr isn't giving me a box to reply to this, so commenting here @BSynRedhat that I use `git commend` when I work so I always only have 1 commit. Its basically a `git ammend` to a commit. You can still see diffs in the comparisons where you see that I force pushed in the PR dialogue.

skrthomas commented on modules/guardrails-orchestrator-hap-scenario.adoc:
```diff
@@ -0,0 +1,138 @@
+----
+$ oc apply -f orchestrator_cm.yaml -n <TEST_NAMESPACE>
+----
+
```
Good catch! Added.

skrthomas commented:
@BSynRedhat when you get a chance, please take a look at the updates from your most recent review. Thank you!
